### ISSUE TITLE
Odd pixels in output of psc_bubble_yz

### ISSUE THREAD

**Issue by jackmatt16** [psc-code/psc#64]
Finding an odd pixelizating effect that seems to correspond to the number of patches - haven't reproduced this for the flatfoil case, but have find it for gpu/cpu in the bubble case - odd right?

![m0002](https://user-images.githubusercontent.com/1091267/64047006-8d8f5280-cb3b-11e9-9c7d-e7b80eb4cf54.png)

**Comment by wrfox** [comment]
Looks like ghost cell problem

**Comment by germasch** [comment]
It definitely doesn't look right. Could you send me, or attach here, the `psc_bubble_yz.cxx` file?

**Comment by jackmatt16** [comment]
This is just in the psc_bubble_yz.cxx in master- no modifications.

**Comment by germasch** [comment]
So I've been tearing my hair out over this to no avail...

I've found one clue, but I actually think i's not likely related. If I look at the error in Gauss's law, which should be satisfied exactly, I see the patches very clearly:

<img width="916" alt="Screen Shot 2019-09-09 at 3 13 04 PM" src="https://user-images.githubusercontent.com/11023174/64559495-6cd1b480-d314-11e9-9de3-90a00f4d2c67.png">

So, to be exact, the charge continuity equation should be satisfied exactly, which in practice means to machine precision. As the E field is updated by Ampere's Law, Gauss's law would be maintained, but in practice, errors from the finite precision charge continuity will accumulate, slowly.

Overall, if I check Gauss's law over time, I see the expected behavior: errors starting at 1e-7, and climbing to 1e-6 and then 1e-5.  However, when plotting the error field, as above, one can see that it's not just random noise but systematic at patch boundaries. But the error still is related to machine precision.

If I run the simulation in double precision, the picture doesn't change fundamentally. Errors start at 1e-15 at eventually grow to 1e-14 or 1e-13 -- still showing that same pattern, though. The original problem, the patch pattern in jx and other quantities, however doesn't change significantly between single and double precision runs. That's why I think the original problem and this pattern in the Gauss law error are probably not actually related.

In fact, I think I spent some days on the Gauss's law issue some years ago, before eventually giving up and concluding (ie, hoping) that it's not a bug in the code. I guess the main reason for that conclusion is that the error introduced by an actual bug, say with ghost cell communications, wouldn't scale with the precision used.

But anyway, it's disconcerting (and for longer runs definitely a reason to do some Marder-correction thing when using single precision, even if it's not related to the issue a hand.)

My attempt at explaining the systematic but machine-precision level error at patch boundaries goes like that: Currents get deposited by following each particles motion and adding little contributions to the current field on the grid. A lot of the time, in particular when there isn't a lot of current in the plasma in a particular direction, one will end up depositing random positive and negative contributions. Which isn't a good thing with limited precision, as the relative error of the result is something like (average absolute error of contributions / magnitude of result), so if the magnitude of the result is small, the relative error becomes large. 

In the interior of a patch, the contributions to a given j on the grid will randomly come from all particles in the vicinity, say going left or right. Near, say, the right border of a patch, though, only particles going to the right will contribute to the rightmost jx location, though -- these are particles that are leaving the patch. So all contributions to jx will be positive. On the patch one over to the right, only particles moving to the left will contribute to the left most jx, making it negative. During the ghost point exchange, currents in the ghost points will be added back to the interior currents of the neighboring patch. So in the end all current contributions on the grid are (should be, anyway) correct. But in practice, I think instead of randomly adding positive and negative contributions, one may end up summing, say, mostly positive contributions first, and then subtracting the accumulated negative contributions afterwards. Which with finite precision isn't quite the same thing, one loses a couple of bits of precision, and that's my best (well, only) guess as to what may be causing the systematic looking errors in Gauss's Law.  Does that logic make sense? I'm not sure it's enough to explain that plot.

But even if it does, I'm afraid the original problem here is another one, and I don't seem to have a handle on it :(.

**Comment by wrfox** [comment]
Hi guys,  couple thoughts and questions

It looks like the effect on the B field is not symmetric as you go around the edge of the plume - it is only pixelated on two of the four quadrants.  Maybe it is a clue?

Second, if we run in double precision, does the pattern in n and T still emerge?    It sounds like the E field errors would be much smaller in that case.

Does it still occur with uniform initial conditions and zero B fields?   Or does it rely on the bubble IC?

Will

**Comment by germasch** [comment]
It's true that there are certain broken symmetries. But so far I still don't really know where I should be looking for the underlying cause of this problem, and without having a clue, the symmetries haven't really helped me, either.

The results in double precision are essentially the same. That's why I don't think that Gauss Law error is related. They're also the same if I run this in quasi-3d (with 2 grid cells in the x direction), or with 2nd order, or if I run the problem in the x-z plan, rather than y-z. Also the same, whether with / without GPUs. So it's a rather universal problem, which otherwise I haven't seen, though that doesn't really mean too much.

**Comment by wrfox** [comment]
The density and temperature look like they are off by a significant factor.  By eye it looks like the density spikes at the patch boundaries by a factor of up to ~2.

How many steps into the sim was Jack’s plot?  Does it look like this in the initial condition or does it develop?

Does it occur for totally uniform IC?   Or just bubble?

Where are particles initialized in each cell?   And are particles initialized in ghost cells (hopefully not...).   Are we double-adding particles at the edges of the patch because of ghost cells?  Is there off-by-one in determining whether the last grid point of a patch is in that patch or the neighbor?

W

**Comment by germasch** [comment]
Yes, the errors that we're talking about are of order unity, very different order of magnitude than the 1e-14 errors in Gauss's Law.

I'm not quite sure about Jack's plot, but the problem shows quickly, 200 or 400 steps into the simulation. The i.c. looks fine. I just tried an all uniform i.c., and the problem still occurs, though it's less pronounced.

Particles are initialized as usual, at cell centers. I don't think it's any of the issue that you mention, though I don' have any better ideas, and it must be something...

**Comment by germasch** [comment]
Alright, so after looking at and overlooking the problem a bunch of times, it turned out that I had essentially commented out `randomize_in_cell` by putting an early return statement into it :(

Fixing the collisions to randomly select pair fixes the spurious pattern, at least in the uniform density test case. Now I've gotta go fix this in the CUDA case, too, which isn't going to be quite as trivial.

**Comment by wrfox** [comment]
Okay, makes sense to me that that would cause the breakage at the ghost cells, because those cells end up with very non-random particle orders owing to the boundary exchange.

The standard randomize in cell swaps each particle with a random partner, and that is basically not threadable - is that the problem with CUDA?

I can picture in my head some shuffling operations which could be threadable, except that they require a few passes through the particle list (or subsidiary order list).     Consider a few passes, where on pass K you swap or don’t swap (50-50 odds) particle N with N+ 2^K.  That should do it and would be threadable.    

Or a hybrid that uses the original algorithm in parallel on small chunks (8,16?) particles, then some of of these random “far” swaps to make sure particles can visit a different spot in the array.   < log_2 (NPPC/16) passes would be needed wheere NPPC is the number of partices in a cell.

Will

**Comment by germasch** [comment]
So what I ended up doing works like this:

The GPU code needs to get particles into sorted-by-cell order first, since normally particles are only sorted by (e.g, 4x4) block. So it previously went over all particles and stored pairs of (cell index, particle index). These pairs are then sorted. That sorted list allows one to find all particles in the first cell, then all particles in the 2nd cell, etc. (It is really just like actually sorting the particles, but in essence, one doesn't move the actual particles, one just ends up with the indices of the particles in a given cell.)

I essentially kept that the same, but made the cell index a float, adding a random number 0<+ x < 1 to each cell index. The pairs are now sorted by the float cell index, meaning that one still gets segments of all particles in a cell, but these particles are now in a random order according to the fractional random number that had been added to their cell index.

I think fundamentally this is sound, ie, truly randomizes, and it adds little cost to what's already being done. There are some potential issues with the limited precision of floating point numbers, so when there are, say, a million cells on a GPU, that takes up 20 bits already, so there are only a few bits left for the random fraction. So this in theory could limit how well the randomization works. Similarly, there can be issue where a particle has cell index 1000000 and the random number happens to be 0.99999, it'll end up with a rounded cell index of 1000001. But I wouldn't really expect any ill effects if 1 in a million times a particle is collided with another particle that's actually in a different cell due to a rounding error.  (This is different from the actual boundary exchange, where similar rounding issues can happen, but there they can lead to losing a particle or worse. I had bugs of that kind before, hopefully I found them all.)

Nevertheless, one thing that could be considered would be using the same method but doing it with all integers and some bit shifting, that would avoid all rounding issues, and also would mean all 32 bits in an int could be used, where as in FP, one is basically limited to the mantissa.
